validator.go

```diff
@@ -27,6 +28,77 @@ type messageValidator struct {
 	useTransportMessageAttributes bool
 }
 
+func (v *messageValidator) getPayloadandAttributes(message *Message) ([]byte, map[string]string, error) {
+	err := v.encoder.VerifyKnownMinorVersion(message.Type, message.DataSchemaVersion)
+	if err != nil {
+		return nil, nil, err
+	}
+	err = v.verifyHeaders(message.Metadata.Headers)
+	if err != nil {
+		return nil, nil, err
+	}
+	schema := v.encoder.EncodeMessageType(message.Type, message.DataSchemaVersion)
+	metaAttrs := MetaAttributes{
+		message.Metadata.Timestamp,
+		message.Metadata.Publisher,
+		message.Metadata.Headers,
+		message.ID,
+		schema,
+		v.currentFormatVersion,
+	}
+	messagePayload, err := v.encoder.EncodeData(message.Data, v.useTransportMessageAttributes, metaAttrs)
+	if err != nil {
+		return nil, nil, err
+	}
+	var attributes map[string]string
+	if v.useTransportMessageAttributes {
+		attributes = v.encodeMetaAttributes(metaAttrs)
+	} else {
+		attributes = message.Metadata.Headers
+	}
+	return messagePayload, attributes, nil
+}
+
+func (v *messageValidator) DeserializeFirehose(line []byte) (*Message, error) {
+	// defer reset of useTransportMessageAttributes
+	defer v.withUseTransportMessageAttributes(v.useTransportMessageAttributes)
+	v.withUseTransportMessageAttributes(false)
+	var messagePayload []byte
+	if v.encoder.IsBinary() {
+		messagePayload = make([]byte, len(line)-8)
+		// TLV format: 8 bytes for size of message, n bytes for the actual message
+		copy(messagePayload, line[8:])
+	} else {
+		messagePayload = make([]byte, len(line)-1)
+		// last char is new line, skip that
+		copy(messagePayload, line[:len(line)-1])
+	}
+	return v.deserialize(messagePayload, map[string]string{}, nil)
+}
+
+func (v *messageValidator) SerializeFirehose(message *Message) ([]byte, error) {
+	// defer reset of useTransportMessageAttributes
+	defer v.withUseTransportMessageAttributes(v.useTransportMessageAttributes)
```

Comment:
@maroux  Not sure if this is thread safe to always use `withUseTransportMessageAttributes` as false in the firehose functions?

Reply:
Doesn't look thread safe yeah.